libcontainer/container_linux.go

```diff
@@ -518,57 +518,107 @@ func (c *Container) shouldSendMountSources() bool {
 		return false
 	}
 
-	// We need to send sources if there are bind-mounts.
+	// We need to send sources if there are non-idmap bind-mounts.
 	for _, m := range c.config.Mounts {
-		if m.IsBind() {
+		if m.IsBindOnly() {
 			return true
 		}
 	}
 
 	return false
 }
 
-func (c *Container) newInitProcess(p *Process, cmd *exec.Cmd, messageSockPair, logFilePair filePair) (*initProcess, error) {
-	cmd.Env = append(cmd.Env, "_LIBCONTAINER_INITTYPE="+string(initStandard))
-	nsMaps := make(map[configs.NamespaceType]string)
-	for _, ns := range c.config.Namespaces {
-		if ns.Path != "" {
-			nsMaps[ns.Type] = ns.Path
+// shouldSendIdmapSources says whether the child process must setup idmap mounts with
+// the mount_setattr already done in the host user namespace.
+func (c *Container) shouldSendIdmapSources() bool {
+	// For the time being we require userns to be in use.
+	if !c.config.Namespaces.Contains(configs.NEWUSER) {
+		return false
+	}
+
+	// nsexec.c mount_setattr() requires CAP_SYS_ADMIN in the initial userns.
```

Comment:
```suggestion
	// nsexec.c mount_setattr() requires CAP_SYS_ADMIN in the runc parent's userns
```

Reply:
@AkihiroSuda it requires it in the initial userns for now (no fs supports idmap mounts inside a userns yet). See the manpage comment: https://git.kernel.org/pub/scm/docs/man-pages/man-pages.git/tree/man2/mount_setattr.2#n638

When that is relaxed, maybe we can simplify this and do the mount inside the userns without the nsexec and fd passing (not sure it will be worth as long as we support older kernels). But for now this seems accurate to me.

What do you think?

Reply:
You need CAP_SYS_ADMIN in the user namespace the filesystem was mounted in and CAP_SYS_ADMIN in the user namespace you're trying to idmap the mount to. And you need CAP_SYS_ADMIN in the owning user namespace of the mount namespace you're currently located in.

For most filesystems the user ns of the filesystem will be init_user_ns. That'll change with tmpfs being able to support idmapped mounts.

So you're both wrong. ;)

Reply:
Added a comment with this, thanks!